extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/health/ReactivePgDataSourceHealthCheck.java

```diff
@@ -0,0 +1,45 @@
+package io.quarkus.reactive.pg.client.runtime.health;
+
+import java.util.concurrent.CompletableFuture;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+import org.eclipse.microprofile.health.HealthCheckResponseBuilder;
+import org.eclipse.microprofile.health.Readiness;
+
+import io.quarkus.arc.Arc;
+import io.vertx.pgclient.PgPool;
+
+@Readiness
+@ApplicationScoped
+public class ReactivePgDataSourceHealthCheck implements HealthCheck {
+    private PgPool pgPool;
+
+    @PostConstruct
+    protected void init() {
+        pgPool = Arc.container().instance(PgPool.class).get();
+    }
+
+    @Override
+    public HealthCheckResponse call() {
+        HealthCheckResponseBuilder builder = HealthCheckResponse.named("Reactive PostgreSQL connection health check").up();
+
+        try {
+            CompletableFuture<Void> databaseConnectionAttempt = new CompletableFuture<>();
+            pgPool.query("SELECT 1", ar -> {
+                if (ar.failed()) {
+                    builder.down();
+                }
+                databaseConnectionAttempt.complete(null);
+            });
+            databaseConnectionAttempt.join();
```

Comment:
Same comments:

* how to handle when there are several Pools
* you need to be sure to not block the IO Thread.

Reply:
@cescoffier thanks for the review.

> Same comments:
> 
> * how to handle when there are several Pools

In the meantime multiple reactive datasources are not handled, once added we will for sure need to handle them the same way we have handled them for the Agroal connection healthcheck. 

> * you need to be sure to not block the IO Thread.

Okay,  are there constructs (a bean that I can inject ) that I can use to check if in IO Thread? 
And if in IO Thread case, what should be done?

Reply:
I think this will always be in the IO thread
Is this still a problem even if we  mark the Health handler as blocking though? https://github.com/quarkusio/quarkus/blob/1.3.0.Final/extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java#L106

Reply:
If it's a problem then I suppose we would need a change to run the health checks (probably all, not just this one) in a different thread pool (which one?)

Reply:
> I think this will always be in the IO thread
> Is this still a problem even if we mark the Health handler as blocking though? https://github.com/quarkusio/quarkus/blob/1.3.0.Final/extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java#L106

Thanks @jmartisk for the clarifications and pointers.

 


> If it's a problem then I suppose we would need a change to run the health checks (probably all, not just this one) in a different thread pool (which one?)

Yes, I think so too. @cescoffier WDYT? Preferably we should do it in another followup PR. I can take it if given enough inputs.

Reply:
+1 for moving this to another PR

Reply:
@cescoffier ping

Reply:
yes, it's a blocking route. Still, a blocking route should not be blocked indefinitely. 
I would recommend adding a timeout (10 seconds) because we may not answer at all. The client (Kubernetes) would get an HTTP timeout, but we would consume and waste one of the threads of the worker thread pool.

Reply:
Where should we ideally process it though? Does the health extension need to maintain its own thread pool or is there a common thread pool that would be appropriate to use?

Reply:
@cescoffier I have added a 10s timeout.

+1 with @jmartisk question up there. Should help with the next step.

Reply:
@cescoffier do you have sometime to have another look and see if all comments have been addressed? Thanks.